Pattern comprehension bug fix

systay opened the pull request.

Commit:
Refactor pattern expression handling

Moves projection creation to the planner instead of doing it when creating pipes

Commit:
Solve pattern comprehension inside complex expressions

Before this commit, these expressions were solved using the rule planner,
which lead to weird behaviour

Mats-SX commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/planner/logical/patternExpressionRewriter.scala:
```diff
@@ -63,8 +67,32 @@ case class patternExpressionRewriter(planArguments: Set[IdName], context: Logica
+
+      // replace pattern comprehension
+      // the contained pattern expression for no further processing
+      // by this tree fold
```
I don't understand this comment. I see it was only copy-pasted from above. Could it be improved or removed? What does it mean?

boggle commented:
I'm +1

Commit:
Fix problems around pattern comprehensions

In some situations, planning of pattern comprehensions did
not take into account already bound variables.

boggle committed:
Rewrite pattern comprehensions to track their outer scope in the original ast

This allows us to compute which variables are actually introduced by the
comprehension (vs which are already bound by the outer scope)
